src/Pirouette/Term/TypeChecker.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+module Pirouette.Term.TypeChecker where
+
+import Data.Bifunctor
+import qualified Data.Map as Map
+import Control.Monad.Except
+import Control.Monad.Reader
+import Pirouette.Term.Syntax.Base hiding (Var, TyVar)
+import Pirouette.Term.Syntax.SystemF
+import Pirouette.Term.Syntax.Subst (shift)
+
+-- | Checks an entire set of declarations.
```

Comment:
Should we also typecheck the main term and have a function such as: (pseudo-code below) 

```haskell
typeCheck :: (PrtReadDefs m) => m ()
typeCheck = do
  allDecls <- prtAllDefs
  mainDef <- prtMain
  case typeCheckDecls decls of
    Left err -> throwError ("Type Errors: " ++ pretty err)
    _ -> return ()
```

Now that I write that, I think that we don't have a type for main; maybe we should add a `Maybe (Type lang)` that represents the type of the main term, if any, to `PrtUnorderedDecls`.

Nevertheless, the reason I'm thinking about a function polymorphic on any `PrtReadDefs` monad is that we could
want to run the type-checker at any point in the pipeline, be it before or after sorting and removing cycles in definitions, for instance.

Reply:
We definitely should! In fact, the problem was that even though the program text has a type for `main`, it was not part of `Program`, so I couldn't check it.